src/parser/jobs/mnk/modules/TwinSnakes.tsx

```diff
@@ -0,0 +1,225 @@
+import {Plural, Trans} from '@lingui/react'
+import React from 'react'
+
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import {getDataBy} from 'data'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+
+import {BuffEvent, CastEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import Checklist, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import Combatants from 'parser/core/modules/Combatants'
+import Enemies from 'parser/core/modules/Enemies'
+import Invulnerability from 'parser/core/modules/Invulnerability'
+import Suggestions, {SEVERITY, Suggestion, TieredSuggestion} from 'parser/core/modules/Suggestions'
+
+import DISPLAY_ORDER from './DISPLAY_ORDER'
+
+const TWIN_SNAKES_CYCLE_BUFFER = 3000
+const TWIN_SNAKES_CYCLE_LENGTH = 6
+
+class TwinState {
+	casts: CastEvent[] = []
+	start: number
+	end?: number
+
+	constructor(timestamp: number) {
+		this.start = timestamp
+	}
+
+	// Mainly here in case we care about oGCDs being unbuffed later
+	public get gcds(): number {
+		return this.casts.filter(event => {
+			const action = getDataBy(ACTIONS, 'id', event.ability.guid) as TODO
+			return action && action.onGcd
+		}).length
+	}
+}
+
+export default class TwinSnakes extends Module {
+	static handle = 'twinsnakes'
+
+	@dependency private checklist!: Checklist
+	@dependency private combatants!: Combatants
+	@dependency private enemies!: Enemies
+	@dependency private invuln!: Invulnerability
+	@dependency private suggestions!: Suggestions
+
+	private history: TwinState[] = []
+	private twinSnake?: TwinState
+
+	// Fury used without TS active
+	private failedFury: number = 0
+
+	// Clipping the duration, or dropping for more than TS itself
+	private earlySnakes: number = 0
+	private lateSnakes: number = 0
+
+	// Separate accounting from the window, to handle counting while TS is down
+	private gcdsSinceTS: number = 0
+
+	protected init() {
+		// Hook all GCDs so we can count GCDs in buff windows
+		this.addHook('cast', {by: 'player'}, this.onCast)
+
+		// This gets weird because, we don't wanna penalise if it was from FPF...
+		this.addHook('applybuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onGain)
+		this.addHook('refreshbuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onRefresh)
+		this.addHook('removebuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onDrop)
+
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onCast(event: CastEvent): void {
+		const action = getDataBy(ACTIONS, 'id', event.ability.guid) as TODO
+
+		if (!action) {
+			return
+		}
+
+		// Only include GCDs
+		if (action.onGcd) {
```

Comment:
I think you should be able to flatten this function by adding another conditional to the above return thing
```
if(!action || !action.onGcd) {
return
}
```